RecoHI/HiEvtPlaneAlgos/interface/HiEvtPlaneFlatten.h

```diff
@@ -37,111 +29,325 @@ class HiEvtPlaneFlatten {
+
+  void fill(double psi, double vtx, int centbin)
   {
     if(fabs(psi)>4 ) return;
-    for(int k = 0; k<hOrder; k++) {
-      double fsin = sin(vorder*(k+1)*psi);
-      double fcos = cos(vorder*(k+1)*psi);
-      int indx = GetCutIndx(centbin,vtx,k);
+    for(int k = 0; k<hOrder_; k++) {
+      double fsin = sin(vorder_*(k+1)*psi);
+      double fcos = cos(vorder_*(k+1)*psi);
+      int indx = getCutIndx(centbin,vtx,k);
       if(indx>=0) {
-	flatX[indx]+=fcos;
-	flatY[indx]+=fsin;
-	++flatCnt[indx];
+	flatX_[indx]+=fcos;
+	flatY_[indx]+=fsin;
+	++flatCnt_[indx];
       }
     }
   }
+  void fillOffset(double s, double c, uint m, double vtx, int centbin)
+  {
+    int indx = getOffsetIndx(centbin,vtx);
+    if(indx>=0) {
+      xoff_[indx]+=c;
+      yoff_[indx]+=s;
+      xyoffmult_[indx]+=m;
+      ++xyoffcnt_[indx];
+    }
+  }
+  void fillPt(double ptval, double vtx, int centbin)
+  {
+    int indx = getOffsetIndx(centbin,vtx);
+    if(indx>=0) {
+      pt_[indx]+=ptval;
+      pt2_[indx]+=ptval*ptval;
+      ++ptcnt_[indx];
+    }
+  }
+
+  void setCaloCentRefBins(const int caloCentRefMinBin, const int caloCentRefMaxBin) {
+    caloCentRefMinBin_ = caloCentRefMinBin;
+    caloCentRefMaxBin_ = caloCentRefMaxBin;
+    caloCentRefVal_ = 1.;
+  }
+
+  void updateEt(double vtx, int centbin) {
+    int refmin = getOffsetIndx(caloCentRefMinBin_,vtx);
+    int refmax = getOffsetIndx(caloCentRefMaxBin_,vtx);
+    caloCentRefVal_ = 0;
+    for(int i = refmin; i<=refmax; i++) {
+      caloCentRefVal_+=getPtDB(i);
+    }
+    caloCentRefVal_/=refmax-refmin+1.;
+  }
 
-  double GetFlatPsi(double psi, double vtx, double cent)
+  double getEtScale(double vtx, int centbin) const {
+    if(caloCentRefMinBin_<0) return 1.;
+    int indx = getOffsetIndx(centbin,vtx);
+    if(caloCentRefVal_==0 || getPtDB(indx)==0) return 1.;
+    return caloCentRefVal_/getPtDB(indx);
+   }
+
+  double getW(double pt, double vtx, int centbin) const
+  {
+    int indx = getOffsetIndx(centbin,vtx);
+    if(indx>=0) {
+      double scale = getEtScale(vtx,centbin);
+      double ptval = getPtDB(indx)*scale;
+      double pt2val = getPt2DB(indx)*pow(scale,2);
+      if(ptval>0) return pt*scale-pt2val/ptval;
+    }
+    return 0.;
+  }
+
+  double getFlatPsi(double psi, double vtx, int centbin) const
   {
     double correction = 0;
-    for(int k = 0; k<hOrder; k++) {
-      int indx = GetCutIndx(cent,vtx,k);
-      correction+=(2./(double)((k+1)*vorder))*(flatXDB[indx]*sin(vorder*(k+1)*psi)-flatYDB[indx]*cos(vorder*(k+1)*psi));
+    for(int k = 0; k<hOrder_; k++) {
+      int indx = getCutIndx(centbin,vtx,k);
+      if(indx>=0) correction+=(2./(double)((k+1)*vorder_))*(flatXDB_[indx]*sin(vorder_*(k+1)*psi)-flatYDB_[indx]*cos(vorder_*(k+1)*psi));
     }
     psi+=correction;
     psi=bounds(psi);
     psi=bounds2(psi);
     return psi;
   }
-  
+
+  void updateEP(double s, double c, double w, uint m,  double vtx, int centbin, bool offset)
```

Comment:
splitting up non-const methods into const and non-const part doesn't solve the problem.
My suggestion in https://github.com/slava77/cmssw/compare/CMSSW_7_5_0_pre4/sign549/pull9000-9022-mo...slava77:CMSSW_7_5_0_pre4/sign549/pull9000-9022-fixMt achieved necessary level of "const-ness".

If you don't like this, you can alternatively define a method ::resetEventVariables and call it every event at the beginning of the ::produce method (or any large scope method where the object of this class type are used).
From the list, it seems like you need to reset soff_, coff_, w_, mult_, caloCentRefVal_